### PR TITLE
North Carolina credit for children

### LINKED ISSUE #2331: North Carolina credit for children
Reference: https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children

### PR DESCRIPTION
**Pull request by jiajiagong0123**
Fixes #2331

### Summary
📱🛠️🌐

Add a new policy parameter `ubi_amount` to the US tax and benefit system. Implement a basic universal basic income (UBI) reform that replaces the standard deduction and personal exemption with a flat monthly payment. Update the `reform_us.yaml` file and the `policyengine-us.ipynb` notebook to demonstrate the UBI reform.

> _`is_valid` checks_
> _refactored for clarity_
> _autumn leaves falling_

### Walkthrough
*  Add a new parameter `child_allowance.amount` to the `parameters.yaml` file, which sets the monthly amount of the child allowance benefit per child ()

### CODE REVIEW

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:1]: need to add separate and widow parameters and include in variable calculation

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+description: Credit amount of the CTC for single
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:1]:
```suggestion
description: North Carolina allows for the following child tax credit for single filers.
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:1]: similar in other yaml files

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      2023-01-01: 50_000
+    amount: 
+      2023-01-01: 0
+
+metadata:
+  type: single_amount
+  threshold_unit: currency-USD
+  amount_unit: currency-USD
+  period: year
+  label: NC CTC for single
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:21]:
```suggestion
  label: North Carolina CTC single filer amount
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:21]: and in other yaml files

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  reference:
+  - title: Credit for Children
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:23]:
```suggestion
  - title: North Carolina Department of Revenue - Credit for Children
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/single.yaml:23]: and other yaml files

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: The north carolina credit for children has been discontinued effective 2018-01-01 - have to reflect that in all yaml files

Comment (PavelMakarchuk) [reply changelog_entry.yaml:1]: replaced by the child deduction

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]: same formatting as single

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/married.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/married.yaml:1]: same formatting as single

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+brackets:
+  - threshold:
+      2015-01-01: 0
+    amount:
+      2015-01-01: 125
+      2018-01-01: 0
+  - threshold:
+      2015-01-01: 32_000
+    amount: 
+      2023-01-01: 100
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:11]: older date written first

**Comment by MaxGhenis** [comment]
to request a review please take this out of draft

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+metadata:
+  type: single_amount
+  threshold_unit: currency-USD
+  amount_unit: currency-USD
+  period: year
+  label: North Carolina CTC head of household filer amount
+  reference:
+  - title: North Carolina Department of Revenue - Credit for Children
+    href: https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:26]: could you add the [legal code](https://www.ncleg.gov/EnactedLegislation/Statutes/HTML/BySection/Chapter_105/GS_105-153.10.html) that states that it was repealed

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:26]: also 2017 [intructions](https://www.ncdor.gov/documents/files/2017-d-401-individual-income-tax-instructions/open)

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:26]: all yaml files

File: policyengine_us/variables/gov/states/nc/tax/income/credits/nc_child_tax_credits.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+from policyengine_us.model_api import *
+
+
+class nc_credit_for_children(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/tax/income/credits/nc_child_tax_credits.py:4]:
```suggestion
class nc_ctc(Variable):
```

File: policyengine_us/variables/gov/states/nc/tax/income/credits/nc_child_tax_credits.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/tax/income/credits/nc_child_tax_credits.py:1]: rename file to nc_ctc.py

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/nc/tax/income/credits/nc_child_tax_credits.py:1]: same for test but .yaml

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+description: North Carolina allows for the following child tax credit for head of household filers.
```
Comment (mzx-4real) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]:
```suggestion
description: North Carolina allows for the following progressive child tax credit for head of household filers.
```

Comment (mzx-4real) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]: also for other parameter yaml files.

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+description: North Carolina allows for the following progressive child tax credit for head of household filers.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]:
```suggestion
description: North Carolina provides the following child tax credit amount for head of household filers.
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]: similar language in all yaml

File: policyengine_us/tests/policy/baseline/gov/states/nc/tax/income/credits/nc_ctc.yaml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  period: 2015
+  input:
+    filing_status: SINGLE
+    state_code: NC
+    adjusted_gross_income: 15_000
+    tax_unit_children: 0
+  output:  
+    nc_ctc: 0
+
+- name: 1 child
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/nc/tax/income/credits/nc_ctc.yaml:11]:
```suggestion
- name: 1 child - single
```

File: policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml (hunk 8 above)
Comment (mzx-4real) [inline policyengine_us/parameters/gov/states/nc/tax/income/credits/ctc/head_of_household.yaml:1]:
```suggestion
description: North Carolina provides for the following child tax credit amount for head of household filers, depending on their federal adjusted gross income.
```
similar language in all yaml

File: policyengine_us/variables/gov/states/nc/tax/income/credits/nc_ctc.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "North Carolina credit for children"
+    definition_period = YEAR
+    unit = USD
+    reference = "https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children"
+    defined_for = StateCode.NC
+
+    def formula(tax_unit, period, parameters):
+        children = tax_unit("tax_unit_children", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/nc/tax/income/credits/nc_ctc.py:14]: need to use `ctc_qualifying_children` instead

File: policyengine_us/variables/gov/states/nc/tax/income/credits/nc_ctc.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "North Carolina credit for children"
+    definition_period = YEAR
+    unit = USD
+    reference = "https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children"
+    defined_for = StateCode.NC
+
+    def formula(tax_unit, period, parameters):
+        ctc_qualifying_children = tax_unit("tax_unit_children", period)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/nc/tax/income/credits/nc_ctc.py:14]:
```suggestion
        ctc_qualifying_children = tax_unit("ctc_qualifying_children", period)
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Feel free to use the subsection symbol as well

**Review by MaxGhenis — APPROVED** [review]

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
I don't understand - [NCDOR says this credit exists](https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children), but this PR sets all amounts to zero starting in 2018?

**Comment by PavelMakarchuk** [comment]
> I don't understand - [NCDOR says this credit exists](https://www.ncdor.gov/taxes-forms/individual-income-tax/credit-children), but this PR sets all amounts to zero starting in 2018?

The link on the bottom of this page leads to the [2017 worksheet](https://www.ncdor.gov/documents/2017-d-401-individual-income-tax-instructions). North Carolina replaced the credit for children starting with the [2018](https://www.ncdor.gov/documents/files/2018-d-401-individual-income-tax-instructions/open) tax as it does not appear in any tax forms any more after [2017](https://www.ncdor.gov/documents/files/2017-d-401-individual-income-tax-instructions/open). The [law](https://casetext.com/statute/general-statutes-of-north-carolina/chapter-105-taxation/subchapter-i-levy-of-taxes/article-4-income-tax/part-2-individual-income-tax/section-105-15310-repealed) has been repealed.

**Review by MaxGhenis — APPROVED** [review]
OK indeed, very strange that it doesn't mention on the webpage that it's been gone for five years. This article confirms that NC used to have a CTC and has since repealed it (an EITC too): https://ncbudget.org/refundable-family-tax-credits-would-make-ncs-tax-system-fairer/

Looks like it's non-refundable:
image

**Pull request merged**